pkg/controller/webapplicationfirewallrule/webapplicationfirewallrule.go

```diff
@@ -0,0 +1,79 @@
+package webapplicationfirewallrule
+
+import (
+	"context"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/pkg/errors"
+	crdsv1alpha1 "github.com/replicatedhq/kubeflare/pkg/apis/crds/v1alpha1"
+	"github.com/replicatedhq/kubeflare/pkg/logger"
+)
+
+func ReconcileWAFRuleInstances(ctx context.Context, instance crdsv1alpha1.WebApplicationFirewallRule, zone *crdsv1alpha1.Zone, cf *cloudflare.API) error {
+	logger.Debug("ReconcileWAFRules for zone")
+
+	zoneID, err := cf.ZoneIDByName(zone.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed to get zone id")
+	}
+
+	existingPackages, err := cf.ListWAFPackages(zoneID)
+	if err != nil {
+		return errors.Wrap(err, "failed to list WAF packages")
+	}
+
+	existingRules := []cloudflare.WAFRule{}
+	for _, currentPackage := range existingPackages {
+		rules, err := cf.ListWAFRules(zoneID, currentPackage.ID)
+		if err != nil {
+			return errors.Wrap(err, "failed to list WAF rules")
+		}
+
+		existingRules = append(existingRules, rules...)
+	}
+
+	desiredRules := []*crdsv1alpha1.WAFRule{}
+	if instance.Spec.Rules != nil {
+		desiredRules = append(desiredRules, instance.Spec.Rules...)
+	}
+
+	rulesToUpdate := []cloudflare.WAFRule{}
+
+	for _, existingRule := range existingRules {
+		found := false
+		for _, desiredRule := range desiredRules {
+			if desiredRule.ID == existingRule.ID {
+				found = true
+				isChanged := false
+
+				if desiredRule.PackageID != "" {
+					if desiredRule.PackageID != existingRule.PackageID {
+						isChanged = true
+						existingRule.PackageID = desiredRule.PackageID
+					}
+				}
+
+				if desiredRule.Mode != existingRule.Mode {
+					isChanged = true
+					existingRule.Mode = desiredRule.Mode
+				}
+
+				if isChanged {
+					rulesToUpdate = append(rulesToUpdate, existingRule)
+				}
+			}
+		}
+		if !found {
+			// TODO
```

Comment:
this is presumably the spot for the 'delete the rule' logic?

do we have some way to only manage rules with a given set of tags, etc?

Reply:
We don't, but I actually think this section doesn't make sense at all as kubeflare doesn't create any of these rules but only modifies them. They're managed by Cloudflare, so they can't be deleted.